### DISCUSSION TITLE
Nixlta Integration

### DISCUSSION THREAD

**Discussion by akmalsoliev** [functime-org/functime#198] [General]
Hello Functime Community,

I recently had a productive conversation with @baggiponte regarding the integration of Nixtla's forecasting technology into the Functime ecosystem. We believe combining Nixtla's state-of-the-art time-series forecasting developments with Functime's efficient use of Polars can yield fantastic results.

I've been exploring the core functionalities and examples, particularly intrigued by Luca's presentation at PyData (link: https://github.com/baggiponte/pydata-global-2023-functime/blob/main/notebooks/scalable-forecasting-with-functime.ipynb).

Here are some insights I gathered:

## Data Transformation
- Functime and Nixtla share a similar approach to data structure, typically organizing data into three columns: an ID Key, a date, and a target column.
- The ID Key uniquely identifies data entries and is indexed in a dictionary or hashmap for efficient categorization. 
Reference: https://github.com/functime-org/functime/blob/ba2348d2bb681f2d33ced280696e1cf97233a09a/functime/base/model.py#L66
- Exogenous variables, or additional features beyond the third column, are accommodated on a per-model basis. 
Reference: https://github.com/functime-org/functime/blob/29b680644e5a04f2d4bca28c201b14725de3569c/functime/forecasting/catboost.py#L26
- When data is split into predictors (X) and target (y), the structure mirrors the above logic with the ID Key remaining prominent.

## Model Operation

Functime models operate with two primary methods: `fit` and `predict`. These methods are intuitively named and straightforward to use, making further explanation here unnecessary.

## Integration Strategy

Nixtla's architecture enables seamless integration, as all models follow a uniform approach featuring key functions like fit, predict, forecast, and cross-validation. My proposal includes streamlining Nixtla's library by inheriting core functionalities into Functime. For example Statsforecast class straightforward application of models without the need for additional features like plotting, saving, or loading.

Models would be directly referenced from Statsforecast for proper acknowledgment of the original creators.

This integration promises to leverage the best aspects of both platforms, enhancing our collective capability in forecasting and data analysis.

**Comment by baggiponte** [comment]
Ciao Akmal, thank you very much for your input. I totally dig this idea. Actually, @topher-lo @ngriffiths13 and I were thinking about something strikingly similar for a while. What we envision for functime is a "bring your own forecaster" approach:

1. Retain the current modular forecaster interface (a model class that contains the `.fit`, `.predict`, `.validate`/`.backtest`, `.score` and `.conformalize` interface, for the most popular libraries
2. a `make_forecaster` method and/or decorator to wrap external classes or your own models (e.g. neural networks).

Thanks to the latest Polars improvements (i.e. zero-copy to/from C-oriented numpy numerical arrays), we believe Polars can be the feature and cross-val engine for most ML tasks. Nixtla brings the best models to the table, as well as the most performant implementations, and I believe I speak as all of functime's maintainers when I say we would love to collaborate!

**Comment by akmalsoliev** [comment]
Ciao Luca, 
Been thinking about the approach and there are certain details that wanted to dive deep in to. Couple of things I wanted to documentate with you:
1. `.fit` and `.predict` are relatively easy, np here.
2. `.validate` would be to create a separate sub sample of train, am I correct on that assumption? 
Is it under models?
4. `.backtest` would be used to test the prediction based on either sliding validation or expanding validation.
https://github.com/functime-org/functime/blob/6bbc79f737bcd9bd2c51a5a26f975e3c5843bbf0/functime/base/forecaster.py#L272
6. `.conformalize` conformal prediction of values?
7. `make_forecaster` generates future values.
https://github.com/functime-org/functime/blob/6bbc79f737bcd9bd2c51a5a26f975e3c5843bbf0/functime/ranges.py#

How do you usually save the models? Or is it more it needs to be reintroduced?

**Reply by baggiponte** [reply]
Ciao! I chatted with @topher-lo about functime internals today and will do a deep dive in these days, so I can give you a proper answer later (will be away for April 25th tho).

> 1. `.fit` and `.predict` are relatively easy, np here.

Yes. This is however the biggest point I have to make.

`functime` has a core `Forecaster` object with a [`.fit`](https://github.com/functime-org/functime/blob/0608c78118b5defd42df73b81288e0e7b32fdb59/functime/base/forecaster.py#L174) and a [`.predict`](https://github.com/functime-org/functime/blob/0608c78118b5defd42df73b81288e0e7b32fdb59/functime/base/forecaster.py#L174) method. The standard steps are performed here (apply transforms like lags etc, which are reverted in predict, etc). To create a new model, you just have to implement the `._fit` method which is then called in the proper `.fit()`. The other methods (backtest, conformalize) come for free.

> 2. `.validate` would be to create a separate sub sample of train, am I correct on that assumption?
>    Is it under models?

My bad, this does not exist.

> 3. `.backtest` would be used to test the prediction based on either sliding validation or expanding validation.
>    https://github.com/functime-org/functime/blob/6bbc79f737bcd9bd2c51a5a26f975e3c5843bbf0/functime/base/forecaster.py#L272
> 4. `.conformalize` conformal prediction of values?

Yes!

> 5. `make_forecaster` generates future values.
>    https://github.com/functime-org/functime/blob/6bbc79f737bcd9bd2c51a5a26f975e3c5843bbf0/functime/ranges.py#

No this is a thing I invented as I was writing to make the custom forecaster, but would just be syntactic sugar. The future values are generated by `make_future_ranges` as you correctly identified.

> How do you usually save the models? Or is it more it needs to be reintroduced?

I don't know - we use cloudpickle somewhere. Some stuff is just persisted in memory. How heavy is a model like N-HiTS?

**Reply by akmalsoliev** [reply]
Ciao grande Luca, 

> functime has a core Forecaster object with a [.fit](https://github.com/functime-org/functime/blob/0608c78118b5defd42df73b81288e0e7b32fdb59/functime/base/forecaster.py#L174) and a [.predict](https://github.com/functime-org/functime/blob/0608c78118b5defd42df73b81288e0e7b32fdb59/functime/base/forecaster.py#L174) method. The standard steps are performed here (apply transforms like lags etc, which are reverted in predict, etc). To create a new model, you just have to implement the ._fit method which is then called in the proper .fit(). The other methods (backtest, conformalize) come for free.

I did not find `._fit` method neither on `forecast.py` nor on `model.py`

> How heavy is a model like N-HiTS? 

Depending on the data size, but, if I remember correctly, the model size of around 20 `unique_id` with around 16 points per uid was around 20mb model. We are talking about Neuralforecast here. 

-------

As mentioned in chat if you're using your own implementation and just importing the models, not the model wrapper, then you're loosing the parallelism on AutoARIMA and other models (right now talking only about Statsforecast), imo auto hyper param search is the biggest W in Nixtlaverse.

**Reply by baggiponte** [reply]
CIao Akmal!

Perhaps wrapping the StatsForecast class makes sense. However, I can't tell right now how we could hook functime feature transform and transform inversion.
